Scripts/Spells: Prayer of Mending

matanshukry opened the pull request:
**Changes proposed:**

-  Added scripts to implement Prayer of Mending behaviour
-  Added sql script to link the c++ class to the relevant spells

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

Closes #26007

**Tests performed:**

(Does it build, tested in-game, etc.)

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:** (add/remove lines as needed)

matanshukry committed:
Scripted priest spell Prayer of Mending

mdX7 commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,171 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+            if (!effectInfo)
+                return;
+
+            Unit *caster = GetCaster();
```
`Unit* caster` instead of `Unit *caster` (no space between type and *)

mdX7 commented:
The inheritance thing feels weird somehow though. Having to use duplicate code is more worse though, so it's fine I guess.

mdX7 commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,171 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                return;
+
+            Unit *caster = GetCaster();
+            if (!caster)
```
caster is never null when a spell is being cast

mdX7 commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,171 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+            {
+                // Caster: the player(priest) that casters the prayer of mending
+                // Owner: the player that currently has the prayer of mending aura on him
+                if (Unit* owner = GetUnitOwner())
```
owner should never be null, otherwise this cannot proc. Caster however can be null

matanshukry commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,171 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                // Owner: the player that currently has the prayer of mending aura on him
+                if (Unit* owner = GetUnitOwner())
+                    if (Unit* caster = GetCaster())
+                        if (Aura* aura = owner->GetAura(SPELL_PRIEST_PRAYER_OF_MENDING_AURA))
```
Which one? there are 3 checks in here.
Are you talking only about the `GetUnitOwner()` ?

mdX7 replied:
only the caster check is required here

matanshukry commented on sql/updates/world/master/2021_02_02_00_world_priest_prayer_of_mending.sql:
```diff
@@ -0,0 +1,5 @@
+DELETE FROM `spell_script_names` WHERE `spell_id` = 33110;
```
Oh yeah I forgot about that, my bad.

matanshukry committed:
- Removed unneeded checks
- Added new rows ids into the delete statement in .sql file

matanshukry committed:
Removed "m" prefix from private variables

mdX7 commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,174 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+            void HandleJump(SpellEffIndex /*effIndex*/)
+            {
+                Unit* caster = GetCaster(); // the one that has the aura
+                assert(caster != null);
```
this will never happen, assert not required

mdX7 commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,174 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                // Caster: the player(priest) that casters the prayer of mending
+                // Owner: the player that currently has the prayer of mending aura on him
+                Unit* owner = GetUnitOwner();
+                assert(owner != null);
```
this will never happen, assert not required

Ovahlord replied:
He should use GetTarget() instead which will return the unit that has the aura.

matanshukry replied:
@Ovahlord What's the difference? the comments on the properties sounds vaguely similar, and using GetUnitOwner() seems to use the correct unit from the tests I've made.

mdX7 commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,174 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                return;
+
+            Unit* caster = GetCaster();
+            assert(caster != null);
```
this will never happen, assert not required

mdX7 commented on sql/updates/world/master/2021_02_02_00_world_priest_prayer_of_mending.sql:
```diff
@@ -0,0 +1,5 @@
+DELETE FROM `spell_script_names` WHERE `spell_id` IN (33110, 33076, 41635, 155793);
```
is 33110 an old spellid related to prayer of mending?

illfated replied:
Exists both on wowhead (not in Classic) & TC ubuntu aowow (3.3.5), not sure of the history
- https://www.wowhead.com/spell=33110/prayer-of-mending
- https://tcubuntu.northeurope.cloudapp.azure.com/aowow/?spell=33110
- https://wow.gamepedia.com/Prayer_of_Mending

illfated replied:
BTW, there used to be a rule that
```sql
DELETE FROM `spell_script_names` WHERE `ScriptName` IN (
```
is the preferred method. Many have debated the validity of that rule.

illfated replied:
Unless there is a good & obvious reason for deleting by spell ID, TC prefers using
```sql
DELETE FROM `spell_script_names` WHERE `ScriptName` IN ('spell_pri_prayer_of_mending', 'spell_pri_prayer_of_mending_aura', 'spell_pri_prayer_of_mending_jump');
```

illfated commented:
old-school title format: **Scripts/Spells: Prayer of Mending**

matanshukry committed:
- Removing asserts
- Removing aura check (inside aura spell proc)
- deleting from spell_script_names by script_names rather than spell_id

illfated commented:
`2021_02_02_00_world_priest_prayer_of_mending.sql`: "No newline at end of file" - add missing NewLine and maybe an empty comment line at the top (`-- `).

matanshukry committed:
- Added empty comment line at the beginning of the sql file
- Added empty new line at the end of the sql file

matanshukry commented:
@illfated Like that?

illfated commented:
Yeah, looks quite OK to me.

F.Y.I. : The SQL value rows indent is not strictly needed.
Maybe it is on par with non-TC general SQL codestyle, IDK.
You can choose to start every line on column 1 if you want.

matanshukry commented:
I remember doing it that way from another issue where you fixed my script and you did it, but I could be wrong.
Anyway if you want me to change it without any indent let me know and I'll do it, otherwise I'll just leave it as it is now.

illfated commented:
Nah, leave it as is for now. I simply meant to inform you for future SQL tickets or PR additions. 🙂

matanshukry commented:
Cool, no more indents from now on.
Thanks :)

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+
+            void OnTargetSelect(std::list<WorldObject*>& targets)
+            {
+                targets.remove_if([](WorldObject* worldObject)
```
This script is incorrect
* (minor) nullptr will be inside the target list
* This is aoe target - you need to pick only one target from the prepopulated list (currently you will cast this on everyone)
* Target is supposed to be a random injured unit (skip full health ones), can hit pets but will always prefer players

matanshukry replied:
> (minor) nullptr will be inside the target list
if you mean before, that is why there's a `!worldObject` check inside.
If you mean after, there shouldn't be since `!worldObject` is part of the return (hence if world object is null, it will reutrn true and will be removed)

Shauren replied:
fuck, i wrote that comment wrong. there was supposed to be "nullptr will never be inside the target list"

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                // Caster: the player(priest) that casters the prayer of mending
+                // Owner: the player that currently has the prayer of mending aura on him
+                Unit* owner = GetUnitOwner();
+                Aura* aura = owner->GetAura(SPELL_PRIEST_PRAYER_OF_MENDING_AURA);
```
"the aura being scripted" is accessible from `GetAura()`

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                    caster->CastSpell(owner, SPELL_PRIEST_PRAYER_OF_MENDING_HEAL);
+
+                    // Only cast jump if stack is higher than 0
+                    if (aura->GetStackAmount() > 1)
+                        owner->CastSpell(owner, SPELL_PRIEST_PRAYER_OF_MENDING_JUMP, true, nullptr, nullptr, caster->GetGUID());
+                    else
+                        aura->Remove();
```
```suggestion
                    if (aura->GetStackAmount() > 1)
                        owner->CastSpell(owner, SPELL_PRIEST_PRAYER_OF_MENDING_JUMP, true, nullptr, nullptr, caster->GetGUID());

                    Remove();
```

You always want to remove the aura from original unit

matanshukry replied:
Jump spell needs the aura to check the stack amount.

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+
+        void HandleEffectDummy(SpellEffIndex effIndex)
+        {
+            SpellEffectInfo const* effectInfo = GetEffectInfo(effIndex);
```
`GetEffectInfo(effIndex)` with the same `effIndex` as script hook argument is equivalent to no-arg `GetEffectInfo()` returning "current effect", it is also guaranteed to never return nullptr

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+            if (!effectInfo)
+                return;
+
+            Unit* target = GetExplTargetUnit();
```
Use `GetHitUnit()` in OnEffect** and **Hit hooks

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                return;
+
+            Unit* caster = GetCaster();
+            CastPrayerOfMendingAura(caster, target, effectInfo->BasePoints);
```
Use `GetEffectValue()`, not `effectInfo->BasePoints`, that gives you the value including all possible modifiers (as a general rule, you should never need to access raw effectInfo->BasePoints in any script)

Shauren commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,168 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+public:
+    bool Load() override
+    {
+        _spellInfoHeal = sSpellMgr->GetSpellInfo(SPELL_PRIEST_PRAYER_OF_MENDING_HEAL, DIFFICULTY_NONE);
```
If you rely on existence of some spell, do the checking logic in Validate() to get startup logs

matanshukry committed:
- Validating heal spell in base class
- Added priority to jump spell

Faq commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1253,28 +1253,73 @@ class spell_pri_prayer_of_mending_jump : public SpellScriptLoader
+            {
+                InjuredPlayers,
+                InjuredPets,
+                FullHealthPlayers,
```
where FullHealthPets? :)

matanshukry replied:
If there woudl be, it would be last.
Can it be though? fully healthed pets without any fully healthed players or injured players?

matanshukry replied:
on second thought: it could be if the pet belongs to the person the aura was on, since he is removed regardless of his health. Will add it.

matanshukry committed:
- Added back include
- Added default case

matanshukry committed:
- Commented out unused argument

illfated commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,214 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+                    {
+                        uint32 chosen = urand(0, targets.size() - 1);
+                        uint32 idx = 0;
+                        targets.remove_if([chosen, &idx](WorldObject* worldObject) { return (chosen != idx++); });
```
Let's hope this line is the last one for now (unused parameter).
```suggestion
                        targets.remove_if([chosen, &idx](WorldObject* /*worldObject*/) { return (chosen != idx++); });
```

matanshukry replied:
yeah I saw that. Was trying to make it show up on my VS with cmake flags but couldn't; just did it manually, sadly.

illfated replied:
I know. It can be frustrating to deal with unused parameters/arguments reported by the checks (Travis & ci/circleci) since they stop checking the rest of the code when one is found.

Shauren replied:
There is a better way to do it
```
WorldObject* selected = Trinity::Containers::SelectRandomContainerElement(targets);
targets.clear();
targets.push_back(selected);
```

matanshukry committed:
Commented out unused parameter

matanshukry committed:
- Added full health pets to priorities
- Seelcting random targets using a different method

mdX7 commented:
just checked on retail, it appears that pets never get PoM unless its directly cast on them.
What I did:
1. My ghoul attacked something to get damage (I was at full health)
2. Priest buffed himself PoM and attacked a different mob to trigger PoM
3. I got PoM at full health, while my ghoul was < 100% hp

matanshukry committed:
- Removed injured pets completely

matanshukry committed:
Changed due to cr on issue:
- Checking for pets and players, regardless of health.
- Players have higher priority

Shauren commented:
I wonder if there a injured rule when in a party/raid group and "whatever" when not

Warpten commented:
Priest main, fairly certain it's supposed to be a smart heal, otherwise it's kind of worthless

matanshukry commented:
I changed it twice already, someone needs to make a decision or do a poll or w/e,
or merge it and create a different issue for wrong behaviour.

Either way currently it's working the way @mdX7 said on his last comment.

mdX7 commented:
so i just quickly checked a log from guild @ hungering destroyer for the first ~16s and it appears that the PoM is NOT having any priority in terms of hp - meaning it will not always jump to the player with the lowest hp.

Also it seems like that its jumping to an injured target (at least targets < 100% health, even if its just 99%). Didnt notice that its jumping to a full hp target yet (e.g. when it jumps away from tank and everyone else has full hp)

So I'd say:
- it will always prioritize players > pets
- injured players > full health players
- it jumps to a random injured player, even if that player has 99% and a different one 20%

What do you think about that @Warpten ?

Fabi commented:
I'm pretty sure that it also jumps to players with full health even if there are players that are < 100% hp

matanshukry committed:
Merge branch 'master' into issue/prayer-of-mending

matanshukry committed:
Changed GetUnitOwner() to GetTarget()

illfated commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,207 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+            void OnTargetSelect(std::list<WorldObject*>& targets)
             {
-                if (Unit* caster = GetOriginalCaster())
+                // find best target
+                // Make sure too exclude the player that had the aura (caster in our case)
```
Cosmetic/grammar comment details:
```suggestion
                // Find the best target.
                // Make sure to exclude player wearing the aura (caster in this case)
```

illfated commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,38 +1125,207 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+
+            void HandleHeal(ProcEventInfo & /*eventInfo*/)
+            {
+                // Caster: the player(priest) that cast the prayer of mending
+                // Target: the player that currently has the prayer of mending aura on him
```
Cosmetic/readability:
```suggestion
                // Caster: player(priest) that cast Prayer of Mending
                // Target: player currently wearing Prayer of Mending aura
```

matanshukry committed:
Update src/server/scripts/Spells/spell_priest.cpp

Co-authored-by: Trond B. Krokli <38162891+illfated@users.noreply.github.com>

matanshukry committed:
Update src/server/scripts/Spells/spell_priest.cpp

Co-authored-by: Trond B. Krokli <38162891+illfated@users.noreply.github.com>

Shauren committed:
Cleanup

illfated commented on src/server/scripts/Spells/spell_priest.cpp:
```diff
@@ -1122,39 +1125,133 @@ class spell_pri_power_word_shield : public SpellScriptLoader
+    void HandleJump(SpellEffIndex /*effIndex*/)
+    {
+        Unit* origCaster = GetOriginalCaster(); // the one that started the prayer of mending chain
+        Unit* target = GetHitUnit(); // the target we decided should jump the aura to
```
This comment needs cleanup, poor sentence order.
```suggestion
        Unit* target = GetHitUnit(); // the target we decided the aura should jump to
```

illfated commented:
Also: `unused variable 'owner'` after the last cleanup.

Shauren committed:
remove unused variable

illfated commented:
Codestyle: `enum PriestSpells` may need to be sorted alphabetically, either in this PR or later.

<details>
<summary>Lines sorted Lexicographically Ascending</summary>

```cpp
enum PriestSpells
{
    SPELL_PRIEST_ABSOLUTION                         = 33167,
    SPELL_PRIEST_ANGELIC_FEATHER_AREATRIGGER        = 158624,
    SPELL_PRIEST_ANGELIC_FEATHER_AURA               = 121557,
    SPELL_PRIEST_ANGELIC_FEATHER_TRIGGER            = 121536,
    SPELL_PRIEST_ARMOR_OF_FAITH                     = 28810,
    SPELL_PRIEST_ATONEMENT                          = 81749,
    SPELL_PRIEST_ATONEMENT_HEAL                     = 81751,
    SPELL_PRIEST_ATONEMENT_TRIGGERED                = 194384,
    SPELL_PRIEST_BLESSED_HEALING                    = 70772,
    SPELL_PRIEST_BODY_AND_SOUL                      = 64129,
    SPELL_PRIEST_BODY_AND_SOUL_DISPEL               = 64136,
    SPELL_PRIEST_BODY_AND_SOUL_SPEED                = 65081,
    SPELL_PRIEST_CURE_DISEASE                       = 528,
    SPELL_PRIEST_DISPEL_MAGIC_FRIENDLY              = 97690,
    SPELL_PRIEST_DISPEL_MAGIC_HOSTILE               = 97691,
    SPELL_PRIEST_DIVINE_AEGIS                       = 47753,
    SPELL_PRIEST_DIVINE_BLESSING                    = 40440,
    SPELL_PRIEST_DIVINE_WRATH                       = 40441,
    SPELL_PRIEST_GLYPH_OF_CIRCLE_OF_HEALING         = 55675,
    SPELL_PRIEST_GLYPH_OF_DISPEL_MAGIC              = 55677,
    SPELL_PRIEST_GLYPH_OF_DISPEL_MAGIC_HEAL         = 56131,
    SPELL_PRIEST_GLYPH_OF_LIGHTWELL                 = 55673,
    SPELL_PRIEST_GLYPH_OF_PRAYER_OF_HEALING_HEAL    = 56161,
    SPELL_PRIEST_GLYPH_OF_SHADOW                    = 107906,
    SPELL_PRIEST_GUARDIAN_SPIRIT_HEAL               = 48153,
    SPELL_PRIEST_ITEM_EFFICIENCY                    = 37595,
    SPELL_PRIEST_LEAP_OF_FAITH                      = 73325,
    SPELL_PRIEST_LEAP_OF_FAITH_EFFECT               = 92832,
    SPELL_PRIEST_LEAP_OF_FAITH_EFFECT_TRIGGER       = 92833,
    SPELL_PRIEST_LEAP_OF_FAITH_TRIGGERED            = 92572,
    SPELL_PRIEST_LEVITATE_EFFECT                    = 111759,
    SPELL_PRIEST_MANA_LEECH_PROC                    = 34650,
    SPELL_PRIEST_ORACULAR_HEAL                      = 26170,
    SPELL_PRIEST_PENANCE_R1                         = 47540,
    SPELL_PRIEST_PENANCE_R1_DAMAGE                  = 47758,
    SPELL_PRIEST_PENANCE_R1_HEAL                    = 47757,
    SPELL_PRIEST_PRAYER_OF_MENDING_AURA             = 41635,
    SPELL_PRIEST_PRAYER_OF_MENDING_HEAL             = 33110,
    SPELL_PRIEST_PRAYER_OF_MENDING_JUMP             = 155793,
    SPELL_PRIEST_REFLECTIVE_SHIELD_R1               = 33201,
    SPELL_PRIEST_REFLECTIVE_SHIELD_TRIGGERED        = 33619,
    SPELL_PRIEST_RENEWED_HOPE                       = 197469,
    SPELL_PRIEST_RENEWED_HOPE_EFFECT                = 197470,
    SPELL_PRIEST_SHADOWFORM_VISUAL_WITHOUT_GLYPH    = 107903,
    SPELL_PRIEST_SHADOWFORM_VISUAL_WITH_GLYPH       = 107904,
    SPELL_PRIEST_SHIELD_DISCIPLINE_ENERGIZE         = 47755,
    SPELL_PRIEST_SHIELD_DISCIPLINE_PASSIVE          = 197045,
    SPELL_PRIEST_SPIRIT_OF_REDEMPTION               = 27827,
    SPELL_PRIEST_STRENGTH_OF_SOUL                   = 197535,
    SPELL_PRIEST_STRENGTH_OF_SOUL_EFFECT            = 197548,
    SPELL_PRIEST_T9_HEALING_2P                      = 67201,
    SPELL_PRIEST_THE_PENITENT_AURA                  = 200347,
    SPELL_PRIEST_TWIN_DISCIPLINES_RANK_1            = 47586,
    SPELL_PRIEST_VAMPIRIC_EMBRACE_HEAL              = 15290,
    SPELL_PRIEST_VAMPIRIC_TOUCH_DISPEL              = 64085,
    SPELL_PRIEST_VOID_SHIELD                        = 199144,
    SPELL_PRIEST_VOID_SHIELD_EFFECT                 = 199145
    SPELL_PRIEST_VOID_SHIELD_EFFECT                 = 199145,
};
```

</details>

matanshukry commented:
> Codestyle: `enum PriestSpells` may need to be sorted alphabetically, either in this PR or later.
> 
> Lines sorted Lexicographically Ascending
> edit: Forgot to remove the red line from PR source view. Removed now.

Didn't know that btw! cool.

illfated commented:
Oh, and BTW, also codestyle/cosmetic: The last comma in the enum list should be removed, even though it does not matter for the compiler, strictly speaking. TC prefers to end the enum list without an ending comma.
